### PR TITLE
feat: add best_block_number call to BlockIdProvider

### LINKED ISSUE #2351: Add `best_number` call to BlockIdProvider
### Describe the feature

we currently use 

https://github.com/paradigmxyz/reth/blob/370c39525a4ae18894e9681f9b6660f9385f76b2/crates/storage/provider/src/traits/block_id.rs#L8-L9

a lot when looking up the best_number like:

https://github.com/paradigmxyz/reth/blob/370c39525a4ae18894e9681f9b6660f9385f76b2/crates/rpc/rpc/src/eth/filter.rs#L219-L220

this currently reads from disk, we can optimize reads by adding dedicated impl for best block number

## TODO
* add extra call like https://github.com/paradigmxyz/reth/blob/370c39525a4ae18894e9681f9b6660f9385f76b2/crates/storage/provider/src/providers/mod.rs#L114-L114

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by chirag-bgh**
Closes #2351

### CODE REVIEW

**Comment by gakonst** [comment]
How does this optimize reads? Isn't it the same as before but behind a helper function?

**Comment by rkrasiuk** [comment]
@gakonst `chain_info` performs one additional lookup for block hash
https://github.com/paradigmxyz/reth/blob/2a8c8d095b7b3e69cba2272b004f486832730ba0/crates/storage/provider/src/providers/mod.rs#L120

File: crates/storage/provider/src/providers/mod.rs [hunk 1]
```diff
@@ -121,6 +121,14 @@ impl<DB: Database> BlockIdProvider for ShareableDatabase<DB> {
         Ok(ChainInfo { best_hash, best_number, last_finalized: None, safe_finalized: None })
     }
 
+    fn best_block_number(&self) -> Result<BlockNumber> {
```
Comment (rkrasiuk) [inline crates/storage/provider/src/providers/mod.rs:124]: call this function in `chain_info` above

**Review by rkrasiuk — CHANGES_REQUESTED** [review]
pls, review the usages of `chain_info`. there are a couple more places where it can be swapped for `best_block_number`

**Review by mattsse — APPROVED** [review]
ty,

lgtm

**Review by gakonst — APPROVED** [review]

**Pull request merged**